Drivers: Uart: Add hardware flow control support for STM32

Benichou34 opened the pull request:
Add option for enable hardware flow control support for STM32.
This option can be individually activated for each uart.

erwango commented on drivers/pinmux/stm32/pinmux_stm32f0.h:
```diff
@@ -10,91 +10,361 @@
+#define STM32F0_PINMUX_FUNC_PA0_USART4_TX \
+	(STM32_PINMUX_ALT_FUNC_4 | STM32_PUSHPULL_PULLUP)
 
-#define STM32F0_PINMUX_FUNC_PB6_USART1_TX \
```
It's ok to reorganize the file, but please do it in a separate commit, this making it hard to review.

Benichou34 replied:
I have created a new pull request for this.

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,6 +25,13 @@ config UART_1
 	  Enable support for USART1 port in the driver.
 	  Say y here if you want to use USART1 device.
 
+config UART_1_HW_FC
```
Please check how we could  introduce dtscts support as a device tree property (uart-has-rtscts) instead of Kconfig.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -685,7 +689,8 @@ static const struct uart_stm32_config uart_stm32_cfg_##name = {		\
 static struct uart_stm32_data uart_stm32_data_##name = {		\
-	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE			\
+	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE,			\
+	.hw_fc = STM32_##name##_HW_FC \
```
You could have used CONFIG_##name##_HW_FC directly here and spare the lines in each device declaration below

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -630,6 +630,10 @@ static int uart_stm32_init(struct device *dev)
 	/* Set the default baudrate */
 	uart_stm32_set_baudrate(dev, data->baud_rate);
 
+	if (data->hw_fc) {
```
A comment would be nice.

Benichou34 commented:
Hardware flow control can be individually activated for each uart in the device tree (hardware-fc-rtscts).

erwango commented on soc/arm/st_stm32/stm32f0/dts_fixup.h:
```diff
@@ -71,6 +72,34 @@
 #define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+#define DT_UART_STM32_USART_2_HARDWARE_FC_RTSCTS	DT_ST_STM32_USART_40004400_HARDWARE_FC_RTSCTS
+
+#define DT_UART_STM32_USART_3_BASE_ADDRESS   DT_ST_STM32_USART_40004800_BASE_ADDRESS
```
New instances addition should be done in a separate commit (ideally a separate PR)

Benichou34 replied:
I separated the modified files into several commits.

erwango commented on dts/bindings/serial/uart.yaml:
```diff
@@ -24,6 +24,11 @@ properties:
+    hardware-fc-rtscts:
+      type: boolean
+      category: optional
+      description: Initial hardware flow control setting for UART
```
IIRC, behavior of optional booleans is as follows:
If available in the node, generated define will be set to true
If not available, generated define will be set to true
So, I'll put the description as follows: "Hardware flow control enabled at boot time".

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -685,7 +690,8 @@ static const struct uart_stm32_config uart_stm32_cfg_##name = {		\
 static struct uart_stm32_data uart_stm32_data_##name = {		\
-	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE			\
+	.baud_rate = DT_UART_STM32_##name##_BAUD_RATE,			\
+	.hw_fc = DT_UART_STM32_##name##_HARDWARE_FC_RTSCTS \
```
Is that field strictly required? We're able to store/load info from IP registers so ideally we don't need to use RAM bits to store the info.

erwango commented on dts/bindings/serial/uart.yaml:
```diff
@@ -24,6 +24,11 @@ properties:
       category: required
       description: Initial baud rate setting for UART
       generation: define
+    hardware-fc-rtscts:
```
Name seems ok to me but I'd like @galak to comment on this.

erwango replied:
@Benichou34 , seems that a `hw-flow-control` is already available for the same purpose, can you check?

Benichou34 replied:
I used the "hw-flow-control" as dts binding keyword.

erwango reviewed (changes requested):
This looks much better using dts!
Please also update uart_stm32_configure/uart_stm32_configure_get apis.

Last, please split your commits into smaller logical entities (yaml, drivers, pins..).

Benichou34 committed:
dts: bindings: stm32-uart: add hw-flow-control binding

Add a DTS binding definition for "hw-flow-control".
The STM32 uart/usart driver shim can use this to enable HW flow control.

Signed-off-by: Benoit Leforestier <benoit.leforestier@gmail.com>

Benichou34 committed:
Drivers: Uart: Add hardware flow control support for STM32

Add option for enable hardware flow control support for STM32.
This option can be individually activated for each uart.

Signed-off-by: Benoit Leforestier <benoit.leforestier@gmail.com>

Benichou34 committed:
dts: fixup: add hardware flow control defines

Add hardware flow control defines in "dts_fixup.h" headers for STM32.

Signed-off-by: Benoit Leforestier <benoit.leforestier@gmail.com>

erwango commented:
Thanks for update!
Meanwhile, another PR has been proposed for the same feature (#13163) by @GeorgeCGV. Can you get along and see which one to keep?

GeorgeCGV commented:
@erwango @Benichou34 
I thought this PR stalled as there was nothing going on for some time. In this PR we miss configuration set, get implementations; and pinmux for F4, and L4.
Apart from that, they are similar.
All of the above can be added here or it can be merged separately, I don't mind.
